### PR TITLE
Prepare libuv integration

### PR DESCRIPTION
**Pull request by Simn**
This sets up the CI for integration with libuv. After merging this, the Haxe compiler is going to depend on libuv.

## Windows (cygwin/mingw)

I have prepared a pseudo cygwin package here: https://github.com/Simn/mingw64-uv/releases/tag/1.32.0

This can simply be unzipped and everything should work. We may or may not look into turning this into an official cygwin package at some point.

## OS X

The homebrew version of libuv can be used.

## Linux

At the moment, libuv has to be built manually.

----

Please check this PR if you're building Haxe yourself so we can find potential problems. If you're not building Haxe, this shouldn't matter to you at all.

There will be a follow-up PR for the API of the new `asys` package, which is then followed by PRs for platform implementations.

### CODE REVIEW

**Comment by RealyUniqueName** [comment]
It spits out some warnings for me on Linux:
```
      ocamlc libs/uv/uv_stubs.o
uv_stubs.c: In function ‘w_fs_lchown’:
uv_stubs.c:287:20: warning: implicit declaration of function ‘uv_fs_lchown’; did you mean ‘uv_fs_fchown’? [-Wimplicit-function-declaration]
   UV_ERROR_CHECK_C(uv_fs_ ## name(Loop_val(loop), Fs_val(req), call, handler), UV_FREE_REQ(Fs_val(req))); \
                    ^
uv_stubs.c:149:15: note: in definition of macro ‘UV_ERROR_CHECK_C’
   int _code = expr; \
               ^~~~
uv_stubs.c:310:2: note: in expansion of macro ‘FS_WRAP’
  FS_WRAP(name, value arg1 COMMA value arg2 COMMA value arg3, CAMLxparam3(arg1, arg2, arg3), , arg1conv(arg1) COMMA arg2conv(arg2) COMMA arg3conv(arg3), handler);
  ^~~~~~~
uv_stubs.c:342:1: note: in expansion of macro ‘FS_WRAP3’
 FS_WRAP3(lchown, String_val, (uv_uid_t)Int_val, (uv_gid_t)Int_val, handle_fs_cb);
 ^~~~~~~~
```
Other than that seems working.

**Comment by Aurel300** [comment]
@RealyUniqueName `uv_fs_lchown` was added to libuv in 1.21.0 (http://docs.libuv.org/en/v1.x/fs.html#c.uv_fs_lchown). Make sure you're not using the libuv1-dev package, which is severely out-dated.

**Comment by RealyUniqueName** [comment]
Yeah, I have it installed.

File: .gitignore [hunk 1]
```diff
@@ -33,6 +33,7 @@
 /libs/xml-light/xml_lexer.ml
 /libs/xml-light/xml_parser.ml
 /libs/xml-light/xml_parser.mli
+/libs/uv/test
```
Comment (Aurel300) [inline .gitignore]: Shouldn't be here anymore.

**Comment by nadako** [comment]
built successfully on my windows ocaml 64-bit setup, but got these warnings:
```
      ocamlc libs/uv/uv_stubs.o
uv_stubs.c: In function 'w_read_start':
uv_stubs.c:661:75: warning: passing argument 3 of 'uv_read_start' from incompatible pointer type [-Wincompatible-pointer-types]
  UV_ERROR_CHECK(uv_read_start(Stream_val(stream), handle_stream_cb_alloc, handle_stream_cb_read));
                                                                           ^
uv_stubs.c:149:15: note: in definition of macro 'UV_ERROR_CHECK_C'
   int _code = expr; \
               ^~~~
uv_stubs.c:661:2: note: in expansion of macro 'UV_ERROR_CHECK'
  UV_ERROR_CHECK(uv_read_start(Stream_val(stream), handle_stream_cb_alloc, handle_stream_cb_read));
  ^
In file included from uv_stubs.c:11:0:
/usr/x86_64-w64-mingw32/sys-root/mingw/include/uv.h:495:15: note: expected 'uv_read_cb {aka void (*)(struct uv_stream_s *, long long int,  const struct uv_buf_t *)}' but argument is of type 'void (*)(uv_stream_t *, long int,  const uv_buf_t *) {aka void (*)(struct uv_stream_s *, long int,  const struct uv_buf_t *)}'
 UV_EXTERN int uv_read_start(uv_stream_t*,
               ^~~~~~~~~~~~~
uv_stubs.c: In function 'w_udp_recv_start':
uv_stubs.c:879:76: warning: passing argument 3 of 'uv_udp_recv_start' from incompatible pointer type [-Wincompatible-pointer-types]
  UV_ERROR_CHECK(uv_udp_recv_start(Udp_val(handle), handle_stream_cb_alloc, handle_udp_cb_recv));
                                                                            ^
uv_stubs.c:149:15: note: in definition of macro 'UV_ERROR_CHECK_C'
   int _code = expr; \
               ^~~~
uv_stubs.c:879:2: note: in expansion of macro 'UV_ERROR_CHECK'
  UV_ERROR_CHECK(uv_udp_recv_start(Udp_val(handle), handle_stream_cb_alloc, handle_udp_cb_recv));
  ^
In file included from uv_stubs.c:11:0:
/usr/x86_64-w64-mingw32/sys-root/mingw/include/uv.h:675:15: note: expected 'uv_udp_recv_cb {aka void (*)(struct uv_udp_s *, long long int,  const struct uv_buf_t *, const struct sockaddr *, unsigned int)}' but argument is of type 'void (*)(uv_udp_t *, long int,  const uv_buf_t *, const struct sockaddr *, unsigned int) {aka void (*)(struct uv_udp_s *, long int,  const struct uv_buf_t *, const struct sockaddr *, unsigned int)}'
 UV_EXTERN int uv_udp_recv_start(uv_udp_t* handle,
               ^~~~~~~~~~~~~~~~~
```

**Comment by Simn** [comment]
@andyli Any concerns with regards to linux packaging?

**Comment by andyli** [comment]
What is the min required libuv version?

Ubuntu xenial provides 1.8.0. If that's not enough, I will have to look into backporting libuv. I'm not sure how easy is that.

**Comment by Simn** [comment]
I'm afraid that's not enough. According to @Aurel300, we depend on the very recent version 1.31 of libuv. I thought we could go for 1.16, but I suppose this doesn't make much of a difference if the package is outdated anyway.

Is there anything we can do to make this easier? Libuv is very easy to build, but I don't know how that translates to packaging.

**Comment by kLabz** [comment]
No issue on arch linux, `libuv-1.31.0-1` is in the official repos and haxe builds and run without issue (except the String.uppercase warnings).

File: libs/uv/uv_stubs.c [hunk 2]
```diff
@@ -4,14 +4,15 @@
...
 #include <caml/memory.h>
 #include <caml/mlvalues.h>
+// #include <caml/minor_gc.h>
 
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <uv.h>
 
 #if (UV_VERSION_HEX < (1 << 16 | 31 << 8))
```
Comment (Aurel300) [inline libs/uv/uv_stubs.c:14]: There is `UV_VERSION_MAJOR`, `UV_VERSION_MINOR`.

Comment (Simn) [reply libs/uv/uv_stubs.c:14]: That wouldn't make the check any simpler.

**Comment by Simn** [comment]
We're hitting some roadblocks with regards to the libuv integration which make it pointless to merge this. I'll keep the branch, but close the PR for now because it doesn't look like we'll be able to solve this in the near future.

The focus will instead be on integrating libuv into HashLink. Once that's working properly, we can revisit the compiler-integration again.

**Pull request closed without merging**